### PR TITLE
feat(components): adds scalar copy button component

### PR DESCRIPTION
**Pull request by antlio**
**Changes**

this pr adds the `ScalarCopyButton` component as same behavior/style are used amongst reference/client and fixes in the meantime some style regarding background inheritance.

| state | preview |
| -------|------|
| before | image |
| after | image |

**Checklist**

I've gone through the following:

- [x] I've added an explanation _why_ this change is needed.
- [x] I've added a changeset (`pnpm changeset`).
- [x] I've added tests for the regression or new feature.
- [ ] I've updated the documentation.

### CODE REVIEW

**Review by hanspagel — COMMENTED** [review]
nice! requires approval by @cameronrohani and @hwkr

**Review by cameronrohani — CHANGES_REQUESTED** [review]
liking it a lot!

can we just remove the border as we chose to not have that on the copy buttons :)

File: packages/api-reference/src/v2/blocks/scalar-request-example-block/components/RequestExample.vue [hunk 1]
```diff
@@ -254,7 +254,7 @@ const id = useId()
     class="request-card dark-mode"
     ref="elem">
     <!-- Header -->
-    <ScalarCardHeader class="pr-0.5">
```
Comment (hwkr) [inline packages/api-reference/src/v2/blocks/scalar-request-example-block/components/RequestExample.vue]: It think this was intentional because of the client selector @cameronrohani?

**Review by hwkr — CHANGES_REQUESTED** [review]
Love the code consolidation 🙏  Thanks so much for taking a swing at this!!

First thing is a bugfix / tweak, it looks like the button doesn't have a background which means it can occlude the code snippet text. I chatted with @cameronrohani and he suggested making it `bg-b-2` to match how they were mocked up for docs.

Google Chrome-2025-07-22-15-09-18@2x


Second thing is for adding it to the component library. Normally a top level component should have a `*.stories.ts` associated with it but in this case I think this component is so closely associated with the code block it it should probably actually just live in the `ScalarCodeBlock` folder and maybe actually be named `ScalarCodeBlockCopy` or something like that. That way you can use the codeblock stories to demonstrate the functionality.

I also have a checklist for adding components to the library, not every component has all these things right now but I'm trying to make sure anything new that's added has everything on the list. I added a few notes specific to this PR


- [x] Exported from `@scalar/components`
- [x] Has JSDocs for all:
  - [x] Components (with examples) (see [`ScalarDropdown.vue`](https://github.com/scalar/scalar/blob/main/packages/components/src/components/ScalarDropdown/ScalarDropdown.vue#L1-L18) for a good example of a component doc block)
  - [x] Props
  - [x] Slots
  - [x] Emits
  - [x] Functions
  - [x] Types
- [ ] Allows overriding of Tailwind classes where applicable (see useBindCx)
- [ ] Has stories showcasing any applicable variants (can use the `ScalarCodeBlock` stories)
- [x] Has unit tests covering any applicable interactions (would be good to test both content being both a string and an object and mouse and keyboard interactions).

Thanks again for working on this! :)

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        :class="{ copied: showCopied }" />
+      <ScalarIconCheck
+        class="check-icon size-4"
+        :class="{ visible: showCopied }" />
+    </ScalarButton>
+  </div>
+</template>
+
+<style scoped>
+.scalar-code-block:hover .scalar-code-copy {
```
Comment (hwkr) [inline packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: This should probably be up in the `ScalarCodeBlock` component

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue [hunk 3]
```diff
@@ -0,0 +1,101 @@
+<script setup lang="ts">
+import { ScalarIconCheck, ScalarIconCopy } from '@scalar/icons'
+import { cva, cx } from '@scalar/use-hooks/useBindCx'
+import { useClipboard } from '@scalar/use-hooks/useClipboard'
+import { computed, ref } from 'vue'
+
+import { ScalarButton } from '../ScalarButton'
+
+const props = withDefaults(
```
Comment (hwkr) [inline packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: Vue encourages using [props destructuring](https://vuejs.org/guide/components/props#reactive-props-destructure) now instead of `withDefault`

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  defineProps<{
+    /** Content to copy to clipboard */
+    content: string | object
+    /** Optional aria label for accessibility */
+    ariaLabel?: string
```
Comment (hwkr) [inline packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: Elsewhere in the component library we just use `label` or `srLabel`. I've stopped writing `ariaLabel` in camelcase because then people use it on actual HTML elements and it doesn't work (it needs to be `aria-label`)

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    /** Size variant */
+    size?: 'sm' | 'md'
+  }>(),
+  {
+    ariaLabel: 'Copy content',
```
Comment (hwkr) [inline packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: Phrasing suggestion

```suggestion
    ariaLabel: 'Copy',
```

**Comment by hwkr** [comment]
A couple of additional accessibility things I thought of while reviewing the code:

1. We probably need some sort of announcement when it's copied for non-sighted users. Could be an `sr-only` block with [`role="alert"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/alert_role) or it could be a visible label with something similar.
2. If we really want to good about our a11y practices we should probably associate the button with the content (code block) that it copies via an [`aria-controls`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-controls) attribute.

File: packages/api-reference/src/v2/blocks/scalar-request-example-block/components/RequestExample.vue (hunk 1 above)
Comment (antlio) [reply packages/api-reference/src/v2/blocks/scalar-request-example-block/components/RequestExample.vue]: i'll keep fixing it in #6310 since its not aligned

**Comment by antlio** [comment]
thank you so much for the thorough review! diving into it!

>  looks like the button doesn't have a background which means it can occlude the code snippet text

went that way to inherit the background instead as we do not use the same one within reference/client cc @cameronrohani however it requires the codeblock to have one set to avoid transparency. should we just set one by default there instead?

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue (hunk 3 above)
Comment (antlio) [reply packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: good catch!

File: packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue (hunk 2 above)
Comment (antlio) [reply packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: yes not against it!

Comment (antlio) [reply packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: actually this is not even enough for code input usage

Comment (antlio) [reply packages/components/src/components/ScalarCopyButton/ScalarCopyButton.vue]: removing copy button from code input from now on as not really accessible and only used in global environment within client

**Review by hwkr — CHANGES_REQUESTED** [review]
Looks good! I pushed some tweaks for the id and stories (`nanoid` isn't stable for SSR I think)

It looks like the padding is breaking the scrollbars on the ResponseBodyRaw and also the button is getting hidden behind the codemirror instance so your can't click it

This PR:
Google Chrome-2025-07-23-17-11-01@2x


Prod:
Google Chrome-2025-07-23-17-11-16@2x

File: packages/api-client/src/views/Request/ResponseSection/ResponseBodyRaw.vue [hunk 6]
```diff
@@ -2,8 +2,7 @@
 import { ScalarCodeBlockCopy } from '@scalar/components'
 import { prettyPrintJson } from '@scalar/oas-utils/helpers'
 import { useCodeMirror, type CodeMirrorLanguage } from '@scalar/use-codemirror'
-import { nanoid } from 'nanoid'
-import { ref, toRef } from 'vue'
+import { ref, toRef, useId } from 'vue'
```
Comment (antlio) [inline packages/api-client/src/views/Request/ResponseSection/ResponseBodyRaw.vue:5]: good call ty!

**Comment by antlio** [comment]
great catch + stacking issue, should be fix now let's go!

**Comment by hwkr** [comment]
> looks good, but the copy button is above the scrollbar, is that what we want?

I think it's okay? cc @cameronrohani

File: packages/use-codemirror/src/hooks/useCodeMirror.ts [hunk 7]
```diff
@@ -296,11 +296,11 @@ function getCodeMirrorExtensions({
     syntaxHighlighting(defaultHighlightStyle, { fallback: true }),
     EditorView.theme({
       '.cm-line': {
-        lineHeight: '20px',
+        lineHeight: '22px',
```
Comment (hwkr) [inline packages/use-codemirror/src/hooks/useCodeMirror.ts:299]: This affects all the codeblocks yeah? Does it have any downstream effects?

Comment (antlio) [reply packages/use-codemirror/src/hooks/useCodeMirror.ts:299]: this should follows recent font size increases, didn't catch any downstream yet

Comment (antlio) [reply packages/use-codemirror/src/hooks/useCodeMirror.ts:299]: and only the codeblock using code mirror (so not in the response example)

**Comment by antlio** [comment]
@hwkr good catch, fixed

**Review by cameronrohani — APPROVED** [review]

File: packages/api-client/src/components/CodeInput/CodeInput.test.ts [hunk 8]
```diff
@@ -159,15 +158,6 @@ describe('CodeInput', () => {
...
 
-  it('copies content when copy button is clicked', async () => {
-    wrapper = createWrapper({ isCopyable: true })
-
-    const copyButton = wrapper.find('.copy-button')
-    await copyButton.trigger('click')
-
-    expect(vi.mocked(useClipboard).mock.results[0]?.value.copyToClipboard).toHaveBeenCalledWith(wrapper.vm.modelValue)
-  })
-
```
Comment (hanspagel) [inline packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: we should keep a test that makes sure there’s a way to copy the content, even if the actual button isn’t part of the component anymore

File: packages/api-reference/src/features/example-responses/ExampleResponses.test.ts [hunk 9]
```diff
@@ -320,33 +320,6 @@ describe('ExampleResponses', () => {
...
-    })
-
-    const copyButton = wrapper.find('.copy-button')
-    expect(copyButton.exists()).toBe(true)
-
-    await copyButton.trigger('click')
-
-    // Since we can't test clipboard directly in jsdom, we can verify the click handler was called
-    expect(wrapper.emitted()).toBeDefined()
-  })
```
Comment (hanspagel) [inline packages/api-reference/src/features/example-responses/ExampleResponses.test.ts]: we should keep a test that makes sure there’s a way to copy the content, even if the actual button isn’t part of the component anymore

and for this component specifically, it would also be good to check **which** example response is copied

File: packages/api-client/src/components/CodeInput/CodeInput.test.ts (hunk 8 above)
Comment (antlio) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: the test has been moved to the scalar code block test

Comment (antlio) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: but as it is not used anymore within code input, any reason why to keep it there?

**Comment by hwkr** [comment]
Looks good! Going to move it to a draft pending release then we can get it in

**Comment by amritk** [comment]
Lotta in depth reviews on this one! Whats the current status?

File: packages/api-client/src/components/CodeInput/CodeInput.test.ts (hunk 8 above)
Comment (hanspagel) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: yeah, we still want to make sure there is _some way to copy the content_ when using the `CodeInput` 

we don’t need to test any edge cases or whatever, just to check that there’s a way to copy the content

**Comment by hanspagel** [comment]
@amritk just one unresolved comment left 👍

File: packages/api-client/src/components/CodeInput/CodeInput.test.ts (hunk 8 above)
Comment (antlio) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: there's no copy button within code input (removed in this PR)

Comment (antlio) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: as only being used for the client environment page - not really accessible in the view that will change and not needed now within collection environment

Comment (hanspagel) [reply packages/api-client/src/components/CodeInput/CodeInput.test.ts:170]: got it!

**Comment by hanspagel** [comment]
Is that just a glitch on my machine? This is from the preview link (and it doesn’t happen on client.scalar.com):

https://ed028e40.client-staging-7m0.pages.dev/workspace/default/request/default

https://github.com/user-attachments/assets/9777054e-b5d8-464c-8653-a38f3e6e753d

**Comment by antlio** [comment]
great catch @hanspagel i missed this - on it!

**Review by hanspagel — APPROVED** [review]
good to go 🫡

**Comment by amritk** [comment]
@hwkr re-review?
